plutus-core/src/Language/PlutusCore/Parser.y

```diff
@@ -190,21 +179,52 @@ getStaticBuiltinName = \case
     "ifThenElse"               -> Just IfThenElse
     _                          -> Nothing
 
-mkBuiltin :: a -> T.Text -> Term TyName Name uni a
-mkBuiltin loc ident = 
-   case getStaticBuiltinName ident of 
-      Just b  -> Builtin loc $ StaticBuiltinName b
-      Nothing -> Builtin loc (DynBuiltinName (DynamicBuiltinName ident))
-
--- FIXME: at this point it would be good to have access to the current
--- dynamic builtin names to check if a builtin with that name exists
--- and issue an error if not.  Unfortunately that involves adding a
--- parameter of type DynamicBuiltinNameMeanings to almost every
--- function in PlutusPrelude.hs, and others elsewhere.  @effectfully
--- says that his future plans may help with this.  In the meantime,
--- you get a scope resolution error during typechecking/execution if
--- it encounters a nonexistent name.  [We could also have a hard-coded
--- list of known dynamic names here, but that might not be ideal.]
+
+--- Parsing built-in types and constants ---
+
+-- | Tags of types in the default universe.
+tagOfTyName :: T.Text -> Maybe Int
+tagOfTyName = \case 
+    "bool"       -> Just $ tagOf DefaultUniBool
+    "bytestring" -> Just $ tagOf DefaultUniByteString
+    "char"       -> Just $ tagOf DefaultUniChar
+    "integer"    -> Just $ tagOf DefaultUniInteger
+    "string"     -> Just $ tagOf DefaultUniString
+    "unit"       -> Just $ tagOf DefaultUniUnit
+    _ -> Nothing
+
+-- | Given a type name, return a type in the (default) universe.
+-- This can fail in two ways: there's not type with that name, or uniAt
+-- fails because it's been given an uknown tag.  In both cases we report
+-- an unknown built-in type.
+getTypeFromTyName :: Closed uni => AlexPosn -> T.Text -> Parse (Some (TypeIn uni))
```

Comment:
It disturbs me that we don't need `DefaultUni <: uni` here.

Reply:
> It disturbs me that we don't need `DefaultUni <: uni` here.

`tagOfTyName` parses a type in every universe, but only correctly handles the default one essentially. In particular, if we have some other universe that also contains `unit`, but with a different integer tag, then `tagOfTyName` won't handle this situation correctly. So we do not need the `DefaultUni <: uni` constraint, because it's essentially hardcoded in the logic.

Reply:
I know that, but it still disturbs me!  This problem should go away when we make the parser parametric over the builtins though, I hope.

Reply:
But it's already parametric over `uni`? Do you mean built-in functions? How are they related?

Reply:
I think I was using "parametric" in a very loose sense.  I just mean that we'll probably be passing the parser (and probably everything else too) some object/structure/whatever that contains all that it needs to know about built-in types and functions. There will presumably be more information about types than what's currently in `uni`, like what the string representations of the type names are and some means of getting the tag from the name. (I haven't thought all this through in detail: I just wanted to deal with low-level stuff about syntax and lexing in this PR before we move on to higher level stuff like that, since that seemed better than having a single giant PR)